Security

ipantazi opened the pull request:
Added Liquibase changesets to insert books, add column deleted to `users`, insert users, create table roles, create table user_roles, insert roles into DB and assign role for existing users.

Added @PreAuthorize with proper role access over the existing endpoints. 

Added the implementation for UserDetailsService interface. 

Added spring-boot-starter-security to the project. 

Added SecurityConfig class with PasswordEncoder and SecurityFilterChain beans. 

Implement UserDetails interface in the User class. 

Added Role entity.

Added RoleRepository.

Added the following fields to the User entity, UserResponseDto and UserRegistrationRequestDto. 

Added methods setRolesToModel and setRolesToDto to UserMapper. 

Added method findByEmail to UserRepository.

Added method initializeRolesCache to UserServiceImpl.

ipantazi committed:
Added Liquibase changesets to insert books, add column deleted to `users`, insert users, create table roles, create table user_roles, insert roles into DB and assign role for existing users.
Added @PreAuthorize with proper role access over the existing endpoints.
Added the implementation for UserDetailsService interface.
Added spring-boot-starter-security to the project.
Added SecurityConfig class with PasswordEncoder and SecurityFilterChain beans.
Implement UserDetails interface in the User class.
Added Role entity.
Added RoleRepository.
Added the following fields to the User entity, UserResponseDto and UserRegistrationRequestDto.
Added methods setRolesToModel and setRolesToDto to UserMapper.
Added method findByEmail to UserRepository.
Added method initializeRolesCache to UserServiceImpl.

liliia-ponomarenko commented on src/main/java/onlinebookstore/dto/user/UserRegistrationRequestDto.java:
```diff
@@ -40,6 +42,10 @@ public record UserRegistrationRequestDto(
-        String shippingAddress
+        String shippingAddress,
+
+        @NotEmpty(message = "Invalid roles. Roles shouldn't be empty or null.")
+        Set<String> roles
+
```
remove this field. It’s  not secure that user can choose a  role during registration

liliia-ponomarenko commented on src/main/java/onlinebookstore/dto/user/UserResponseDto.java:
```diff
@@ -1,10 +1,14 @@
+    private String firstName;
+    private String lastName;
+    private String shippingAddress;
+    private Set<String> roles;
```
```suggestion
```

liliia-ponomarenko commented on src/main/java/onlinebookstore/mapper/UserMapper.java:
```diff
@@ -1,16 +1,55 @@
     User toModel(UserRegistrationRequestDto userRegistrationRequestDto);
 
+    @AfterMapping
+    default void setRolesToModel(
```
you can remove it

liliia-ponomarenko commented on src/main/java/onlinebookstore/model/User.java:
```diff
@@ -31,4 +47,49 @@ public class User {
+    @ToString.Exclude
+    @HashCodeExclude
+    @EqualsExclude
+    @Column(nullable = false)
```
```suggestion
```

liliia-ponomarenko commented on src/main/java/onlinebookstore/service/user/UserServiceImpl.java:
```diff
@@ -1,28 +1,52 @@
+    @PostConstruct
+    public void initializeRolesCache() {
+        List<Role> allRolesFromDb = roleRepository.findAll();
+        if (allRolesFromDb.isEmpty()) {
+            throw new InvalidRoleException("No roles found in the database");
+        }
```
is it possible?

liliia-ponomarenko commented on src/main/java/onlinebookstore/service/user/UserServiceImpl.java:
```diff
@@ -1,28 +1,52 @@
 
     @Override
-    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+    public UserResponseDto register(UserRegistrationRequestDto userRegistrationDto)
```
provide default USER role

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s  improve your solution ;)

ipantazi committed:
Deleted field roles from class UserResponseDto, UserRegistrationRequestDto record.
Deleted methods setRolesToModel and setRolesToDto from interface UserMapper.
Deleted annotation @Column(nullable = false) from class User.
Deleted class InvalidRoleException, method handleInvalidRoleException from class CustomGlobalExceptionHandler and check from method initializeRolesCache in class UserServiceImpl.
Applying a default role in class UserServiceImpl.

ahoienko reviewed (approved):
Good job!
